File: worker/github.mjs
import portfolio from '../.generated/portfolio.json' with {type: 'json'};
import {HttpError, nowSeconds} from './security.mjs';

const color = value => /^#[0-9a-f]{6}$/i.test(value || '') ? value : '#8b949e';
const text = value => typeof value === 'string' ? value : '';
const count = value => Number.isSafeInteger(value) && value >= 0 ? value : 0;
const publicUrl = value => {
  try { const url = new URL(value); return url.protocol === 'https:' ? url.href : ''; } catch { return ''; }
};

function normalize(user) {
  const repos = Array.isArray(user?.publicRepos?.nodes) ? user.publicRepos.nodes : [];
  const contributions = user?.contributionsCollection || {};
  const days = (contributions.contributionCalendar?.weeks || []).flatMap(week => week.contributionDays || [])
    .map(day => ({date: text(day.date), count: count(day.contributionCount)}));
  const peak = Math.max(0, ...days.map(day => day.count));
  for (const day of days) {
    const ratio = peak ? day.count / peak : 0;
    day.level = day.count === 0 ? 0 : ratio > .75 ? 4 : ratio > .5 ? 3 : ratio > .25 ? 2 : 1;
  }
  const languages = new Map();
  for (const repo of repos) {
    const lang = repo.primaryLanguage;
    if (lang?.name) languages.set(lang.name, {name: text(lang.name).slice(0, 80), color: color(lang.color), count: (languages.get(lang.name)?.count || 0) + 1});
  }
  const langTotal = [...languages.values()].reduce((sum, item) => sum + item.count, 0) || 1;
  const languageList = [...languages.values()].map(item => ({...item, pct: Math.round(item.count * 1000 / langTotal) / 10}))
    .sort((a, b) => b.count - a.count).slice(0, 8);
  const featured = repos.filter(repo => !repo.isFork).sort((a, b) => count(b.stargazerCount) - count(a.stargazerCount)).slice(0, 6).map(repo => ({
    name: text(repo.name).slice(0, 100), description: text(repo.description).slice(0, 500),
    url: publicUrl(repo.url), homepage: publicUrl(repo.homepageUrl), stars: count(repo.stargazerCount),
    forks: count(repo.forkCount), language: text(repo.primaryLanguage?.name).slice(0, 80), color: color(repo.primaryLanguage?.color),
    topics: (repo.repositoryTopics?.nodes || []).map(node => text(node.topic?.name).slice(0, 50)).filter(Boolean).slice(0, 6),
    pushed_at: text(repo.pushedAt).slice(0, 10),
  })).filter(repo => repo.url);
  return {
    configured: true,
    profile: {login: text(user.login), name: text(user.name), avatar: publicUrl(user.avatarUrl), bio: text(user.bio)},
    totals: {public_repos: count(user.publicRepos?.totalCount), private_repos: count(user.privateRepos?.totalCount),
      stars: repos.reduce((sum, repo) => sum + count(repo.stargazerCount), 0), followers: count(user.followers?.totalCount),
      contributions: count(contributions.contributionCalendar?.totalContributions)},
    activity: {commits: count(contributions.totalCommitContributions), prs: count(contributions.totalPullRequestContributions), issues: count(contributions.totalIssueContributions)},
    languages: languageList, featured,
    calendar: {total: count(contributions.contributionCalendar?.totalContributions), peak, weeks: Array.from({length: Math.ceil(days.length / 7)}, (_, i) => days.slice(i * 7, i * 7 + 7))},
    fetched_at: new Date().toISOString().replace(/\.\d{3}Z$/, 'Z'),
  };
}

function merge(primary, secondary) {
  const merged = structuredClone(primary);
  merged.followers.totalCount = count(primary.followers?.totalCount) + count(secondary.followers?.totalCount);
  const firstDays = new Map((primary.contributionsCollection?.contributionCalendar?.weeks || []).flatMap(w => w.contributionDays || []).map(d => [d.date, count(d.contributionCount)]));
  const secondDays = new Map((secondary.contributionsCollection?.contributionCalendar?.weeks || []).flatMap(w => w.contributionDays || []).map(d => [d.date, count(d.contributionCount)]));
  const dates = [...new Set([...firstDays.keys(), ...secondDays.keys()])].sort();
  const contributionDays = dates.map(date => ({date, contributionCount: (firstDays.get(date) || 0) + (secondDays.get(date) || 0)}));
  const a = primary.contributionsCollection || {}, b = secondary.contributionsCollection || {};
  merged.contributionsCollection = {
    totalCommitContributions: count(a.totalCommitContributions) + count(b.totalCommitContributions),
    totalPullRequestContributions: count(a.totalPullRequestContributions) + count(b.totalPullRequestContributions),
    totalIssueContributions: count(a.totalIssueContributions) + count(b.totalIssueContributions),
    contributionCalendar: {totalContributions: contributionDays.reduce((sum, day) => sum + day.contributionCount, 0), weeks: [{contributionDays}]},
  };
  merged.publicRepos = {totalCount: count(primary.publicRepos?.totalCount) + count(secondary.publicRepos?.totalCount), nodes: [...(primary.publicRepos?.nodes || []), ...(secondary.publicRepos?.nodes || [])]};
  merged.privateRepos = {totalCount: count(primary.privateRepos?.totalCount) + count(secondary.privateRepos?.totalCount)};
  return merged;
}

export class GitHubService {
  constructor(store, env, http) { this.store = store; this.env = env; this.http = http; }
  async fetchUser(username, token) {
    if (!username || !token) return null;
    const response = await this.http('https://api.github.com/graphql', {
      method: 'POST', redirect: 'error', signal: AbortSignal.timeout(10000),
      headers: {'Authorization': `bearer ${token}`, 'Content-Type': 'application/json', 'User-Agent': 'sanket-portfolio'},
      body: JSON.stringify({query: portfolio.githubQuery, variables: {login: username}}),
    });
    if (!response.ok) throw new HttpError(502, 'GitHub activity is temporarily unavailable.');
    const body = await response.json();
    if (body.errors || !body.data?.user) throw new HttpError(502, 'GitHub activity is temporarily unavailable.');
    return body.data.user;
  }
  async get() {
    const now = nowSeconds();
    const cached = await this.store.cache('github:v1');
    if (cached && cached.expires > now) return JSON.parse(cached.value);
    if (!this.env.GITHUB_TOKEN || !this.env.GITHUB_USERNAME) return portfolio.fallback;
    let primary = await this.fetchUser(this.env.GITHUB_USERNAME, this.env.GITHUB_TOKEN);
    if (this.env.GITHUB_USERNAME_2 && this.env.GITHUB_TOKEN_2) {
      try { const secondary = await this.fetchUser(this.env.GITHUB_USERNAME_2, this.env.GITHUB_TOKEN_2); if (secondary) primary = merge(primary, secondary); } catch { /* primary remains useful */ }
    }
    const result = normalize(primary);
    await this.store.putCache('github:v1', result, now + 3600);
    return result;
  }
}

File: services/github.py
"""GitHub integration — live stats via the GraphQL API.

Design notes:
  * One GraphQL request per cache window. Results are memoised in-process for
    ``GITHUB_CACHE_TTL_SECONDS`` so page renders cost nothing after the first.
  * Every failure mode is swallowed and reported as ``None``. The site must
    never break because GitHub is unreachable or the token is missing/expired.
  * Only public repository metadata is queried. Private repository access is
    unnecessary and must not be granted to the portfolio token.
"""
import logging
import time
from datetime import datetime, timezone

import requests

log = logging.getLogger(__name__)

GRAPHQL_URL = "https://api.github.com/graphql"
REQUEST_TIMEOUT = 10

# How many public repos to pull in the featured grid.
FEATURED_LIMIT = 6

_QUERY = """
query($login: String!) {
  user(login: $login) {
    login
    name
    avatarUrl
    bio
    followers { totalCount }
    contributionsCollection {
      totalCommitContributions
      totalPullRequestContributions
      totalIssueContributions
      contributionCalendar {
        totalContributions
        weeks {
          contributionDays { date contributionCount }
        }
      }
    }
    publicRepos: repositories(
      privacy: PUBLIC
      first: 100
      orderBy: { field: STARGAZERS, direction: DESC }
      ownerAffiliations: OWNER
    ) {
      totalCount
      nodes {
        name
        description
        url
        homepageUrl
        stargazerCount
        forkCount
        isFork
        pushedAt
        primaryLanguage { name color }
        repositoryTopics(first: 6) { nodes { topic { name } } }
      }
    }
  }
}
"""

# {data, ts} — ts is a monotonic timestamp of the last successful fetch.
_cache = {"data": None, "ts": 0.0}


def _level(count: int, peak: int) -> int:
    """Map a daily contribution count to a 0-4 heatmap level."""
    if count <= 0:
        return 0
    if peak <= 0:
        return 1
    ratio = count / peak
    if ratio > 0.75:
        return 4
    if ratio > 0.5:
        return 3
    if ratio > 0.25:
        return 2
    return 1


def _language_breakdown(repo_nodes):
    """Count primary languages across public repos into a sorted percentage list."""
    counts = {}
    colors = {}
    for repo in repo_nodes:
        lang = repo.get("primaryLanguage")
        if not lang:
            continue
        name = lang["name"]
        counts[name] = counts.get(name, 0) + 1
        colors[name] = lang.get("color") or "#8b949e"

    total = sum(counts.values()) or 1
    breakdown = [
        {
            "name": name,
            "count": count,
            "color": colors[name],
            "pct": round(count * 100 / total, 1),
        }
        for name, count in counts.items()
    ]
    breakdown.sort(key=lambda item: item["count"], reverse=True)
    return breakdown[:8]


def _normalize(user: dict) -> dict:
    """Turn the raw GraphQL response into a stable, template-friendly shape."""
    repos = user["publicRepos"]["nodes"]
    contributions = user["contributionsCollection"]

    days = []
    for week in contributions["contributionCalendar"]["weeks"]:
        for day in week["contributionDays"]:
            days.append(
                {"date": day["date"], "count": day["contributionCount"]}
            )

    peak = max((day["count"] for day in days), default=0)
    for day in days:
        day["level"] = _level(day["count"], peak)

    # Chunk back into Sunday-aligned weeks for a 7-row heatmap grid.
    weeks = [days[i:i + 7] for i in range(0, len(days), 7)]

    candidates = [r for r in repos if not r.get("isFork")]
    featured = sorted(
        candidates, key=lambda r: r.get("stargazerCount", 0), reverse=True
    )[:FEATURED_LIMIT]

    return {
        "configured": True,
        "profile": {
            "login": user.get("login"),
            "name": user.get("name"),
            "avatar": user.get("avatarUrl"),
            "bio": user.get("bio"),
        },
        "totals": {
            "public_repos": user["publicRepos"]["totalCount"],
            "stars": sum(r.get("stargazerCount", 0) for r in repos),
            "followers": user["followers"]["totalCount"],
            "contributions": contributions["contributionCalendar"]["totalContributions"],
        },
        "activity": {
            "commits": contributions["totalCommitContributions"],
            "prs": contributions["totalPullRequestContributions"],
            "issues": contributions["totalIssueContributions"],
        },
        "languages": _language_breakdown(repos),
        "featured": [
            {
                "name": r["name"],
                "description": r.get("description") or "",
                "url": r["url"],
                "homepage": r.get("homepageUrl") or "",
                "stars": r.get("stargazerCount", 0),
                "forks": r.get("forkCount", 0),
                "language": (r.get("primaryLanguage") or {}).get("name", ""),
                "color": (r.get("primaryLanguage") or {}).get("color") or "#8b949e",
                "topics": [t["topic"]["name"] for t in r.get("repositoryTopics", {}).get("nodes", [])],
                "pushed_at": (r.get("pushedAt") or "")[:10],
            }
            for r in featured
        ],
        "calendar": {
            "total": contributions["contributionCalendar"]["totalContributions"],
            "peak": peak,
            "weeks": weeks,
        },
        "fetched_at": datetime.now(timezone.utc).isoformat(timespec="seconds"),
    }


def _fetch_user(username: str, token: str):
    """Raw GraphQL fetch for one account. Returns the ``user`` dict or None."""
    if not token or not username:
        return None
    try:
        response = requests.post(
            GRAPHQL_URL,
            headers={
                "Authorization": f"bearer {token}",
                "Content-Type": "application/json",
            },
            json={"query": _QUERY, "variables": {"login": username}},
            timeout=REQUEST_TIMEOUT,
        )
        response.raise_for_status()
        payload = response.json()
    except Exception:
        log.warning("GitHub GraphQL request failed for configured account")
        return None

    if payload.get("errors"):
        log.warning("GitHub GraphQL returned a provider error")
        return None

    user = (payload.get("data") or {}).get("user")
    if not user:
        log.warning("Configured GitHub account was not found")
    return user


def _merge_users(primary: dict, secondary: dict) -> dict:
    """Combine two accounts' raw GraphQL ``user`` payloads into one, so the
    dashboard can show total contributions/repos/stars across both."""
    merged = dict(primary)

    merged["followers"] = {
        "totalCount": primary["followers"]["totalCount"] + secondary["followers"]["totalCount"]
    }

    def _days(user):
        out = {}
        for week in user["contributionsCollection"]["contributionCalendar"]["weeks"]:
            for day in week["contributionDays"]:
                out[day["date"]] = day["contributionCount"]
        return out

    days1, days2 = _days(primary), _days(secondary)
    all_dates = sorted(set(days1) | set(days2))
    merged_days = [
        {"date": d, "contributionCount": days1.get(d, 0) + days2.get(d, 0)}
        for d in all_dates
    ]

    c1, c2 = primary["contributionsCollection"], secondary["contributionsCollection"]
    merged["contributionsCollection"] = {
        "totalCommitContributions": c1["totalCommitContributions"] + c2["totalCommitContributions"],
        "totalPullRequestContributions": c1["totalPullRequestContributions"] + c2["totalPullRequestContributions"],
        "totalIssueContributions": c1["totalIssueContributions"] + c2["totalIssueContributions"],
        "contributionCalendar": {
            "totalContributions": sum(d["contributionCount"] for d in merged_days),
            "weeks": [{"contributionDays": merged_days}],
        },
    }

    merged["publicRepos"] = {
        "totalCount": primary["publicRepos"]["totalCount"] + secondary["publicRepos"]["totalCount"],
        "nodes": primary["publicRepos"]["nodes"] + secondary["publicRepos"]["nodes"],
    }
    return merged


def get_github_stats(
    username: str, token: str, ttl: int = 3600, username2: str = "", token2: str = ""
):
    """Return normalized GitHub stats, memoised for ``ttl`` seconds.

    When ``username2``/``token2`` are also configured, their contributions,
    repos and stars are merged into the same dashboard. If the second account
    isn't configured, or its fetch fails, this silently falls back to the
    primary account alone — the page never depends on both being present.

    Returns ``None`` when the primary account has no token or its request
    fails — callers should fall back to curated content in that case.
    """
    now = time.monotonic()
    if _cache["data"] is not None and (now - _cache["ts"]) < ttl:
        return _cache["data"]

    user = _fetch_user(username, token)
    if not user:
        return None

    if username2 and token2:
        secondary = _fetch_user(username2, token2)
        if secondary:
            user = _merge_users(user, secondary)

    try:
        stats = _normalize(user)
    except Exception:
        log.warning("Failed to normalize GitHub payload")
        return None

    _cache["data"] = stats
    _cache["ts"] = now
    return stats


def get_github_payload(config, fallback_repos=None):
    """Convenience wrapper for blueprints/templates.

    Always returns a dict. When live data is unavailable it returns a
    ``configured: False`` payload carrying the curated fallback repos so the
    GitHub section renders something meaningful either way.
    """
    stats = get_github_stats(
        config.get("GITHUB_USERNAME", ""),
        config.get("GITHUB_TOKEN", ""),
        config.get("GITHUB_CACHE_TTL_SECONDS", 3600),
        config.get("GITHUB_USERNAME_2", ""),
        config.get("GITHUB_TOKEN_2", ""),
    )
    if stats:
        return stats

    return {
        "configured": False,
        "profile": {"login": config.get("GITHUB_USERNAME", "")},
        "totals": {},
        "activity": {},
        "languages": [],
        "featured": fallback_repos or [],
        "calendar": {"total": 0, "peak": 0, "weeks": []},
        "fetched_at": None,
    }
